packages/lib/src/select.ts

```diff
@@ -12,40 +16,118 @@ function promptFromNaming(config: Naming): string | undefined {
 	return config.itemName ? `Select ${indefiniteArticleFor(config.itemName)} ${config.itemName}.` : undefined
 }
 
-export async function selectGeneric<ID, L>(command: SmartThingsCommandInterface, config: SelectingConfig<L>,
-		preselectedId: ID | undefined, listItems: ListDataFunction<L>,
-		getIdFromUser: IdRetrievalFunction<ID, L>, promptMessage?: string, autoChoose = false): Promise<ID> {
-	if (preselectedId) {
-		return preselectedId
-	}
+export interface PromptUserOptions<L, ID = string> {
+	/**
+	 * A function that returns the list of items to display.
+	 */
+	listItems: ListDataFunction<L>
+
+	/**
+	 * The method that will prompt the user for an item from the list. The default is
+	 * `stringGetIdFromUser` which is usually the right option unless `ID` is not `string`.
+	 */
+	getIdFromUser?: IdRetrievalFunction<ID, L>
+
+	/**
+	 * A message to display to the user when prompting for their choice.
+	 */
+	promptMessage?: string
+
+	/**
+	 * Set this to true if you want to automatically choose for the user if `listItems` returns
+	 * exactly one item.
+	 *
+	 * default: false
+	 */
+	autoChoose?: boolean
+}
 
-	const items = await listItems()
-	if (autoChoose && items.length === 1) {
+/**
+ * Display a list of items and ask the user to choose one. You probably want to use
+ * `selectFromList` instead which also allows for a preselected id (usually from the command line).
+ *
+ * @param command The command that is requesting the selection.
+ * @param config Configuration for primary key, sorting key and fields to display when asking the user for a selection.
+ * @param options More parameters bundled in an object for readability.
+ * @returns Selected id if one was chosen. Logs message if no items are found and exits.
+ */
+export async function promptUser<L, ID = string>(command: SmartThingsCommandInterface,
+		config: SelectingConfig<L>, options: PromptUserOptions<L, ID>): Promise<ID> {
+	const items = await options.listItems()
+	if (options.autoChoose && items.length === 1) {
 		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
 		// @ts-ignore Typescript can't check that run-time variable `config.primaryKeyName` exists.
-		return items[0][config.primaryKeyName]
+		return items[0][config.primaryKeyName] as ID
 	}
 	const list = await outputList(command, config, async () => items, true, true)
 	if (list.length === 0) {
-		// Nothing was found; user was already notified.
+		// Nothing was found; user was already notified by `outputList` above.
 		command.exit(0)
```

Comment:
Would it be better to let the caller decide what they want to do next and instead just return a signal that the list was empty? This causes an ExitError from oclif (displayed red to the user) which may not be desirable in some cases.

Reply:
I think that's just a bug or a debug feature of oclif and is rather inconsistently applied. It doesn't happen when running a command that isn't from the edge plugin or when using released version of the CLI with the built-in edge plugin:

```
MNRSTENERSENM1 dev_pat ➜  bin ~/bin/st-beta.0 edge:channels:unassign           
┌───┬─────────────┬──────────────────────────────────────┐
│ # │ Name        │ Channel Id                           │
├───┼─────────────┼──────────────────────────────────────┤
│ 1 │ Channel Dos │ xxx │
│ 2 │ Channel Uno │ yyy │
└───┴─────────────┴──────────────────────────────────────┘
? Select a channel for the driver. 1
no drivers found
```

I don't see any use-cases for doing anything but exiting in this case. If we ever have one, we can always add an option that disables the exit (but that is off by default). It has come up that we might want to allow other options in the list so there is room for improvement in this method for sure but I think that's another issue. (One that will be easier to implement with this new `options` parameter.)

(I see what you are talking about if I run the CLI with the edge plugin installed via `plugins:link`. This is what happens when I run with the plugin linked:

```
MNRSTENERSENM1 dev_pat ➜  examples st edge:channels:unassign 
┌───┬─────────────┬──────────────────────────────────────┐
│ # │ Name        │ Channel Id                           │
├───┼─────────────┼──────────────────────────────────────┤
│ 1 │ Channel Dos │ 11ea5221-0a28-4679-9f08-3ad689ca4e78 │
│ 2 │ Channel Uno │ 103b5928-8c39-4cc2-98ca-d55b3ee1cf68 │
└───┴─────────────┴──────────────────────────────────────┘
? Select a channel for the driver. 1
no drivers found
 ›   Error: EEXIT: 0
 ›   Code: EEXIT
```
)

Reply:
I did not realize the visible error was only happening in select cases. I agree with leaving it until the need arises.